src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1202,49 +1313,82 @@ CHIP_ERROR CASESession::SendSigma2()
+                                                  sizeof(uint16_t),                    // responderSessionId
+                                                  kP256_PublicKey_Length,              // responderEphPubKey
+                                                  input.encrypted2Length,              // encrypted2
+                                                  SessionParameters::kEstimatedTLVSize // responderSessionParams
+    );
+
+    // the PacketBufferHandler should be empty
+    VerifyOrReturnError(msg_R2.IsNull(), CHIP_ERROR_INCORRECT_STATE);
+
+    msg_R2 = System::PacketBufferHandle::New(data_len);
     VerifyOrReturnError(!msg_R2.IsNull(), CHIP_ERROR_NO_MEMORY);
 
     System::PacketBufferTLVWriter tlvWriterMsg2;
-    outerContainerType = TLV::kTLVType_NotSpecified;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
 
     tlvWriterMsg2.Init(std::move(msg_R2));
     ReturnErrorOnFailure(tlvWriterMsg2.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outerContainerType));
-    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(1), &msg_rand[0], sizeof(msg_rand)));
-    ReturnErrorOnFailure(tlvWriterMsg2.Put(TLV::ContextTag(2), GetLocalSessionId().Value()));
-    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(3), mEphemeralKey->Pubkey(),
-                                                static_cast<uint32_t>(mEphemeralKey->Pubkey().Length())));
-    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(),
-                                                static_cast<uint32_t>(msg_r2_signed_enc_len + CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES)));
 
-    VerifyOrReturnError(mLocalMRPConfig.HasValue(), CHIP_ERROR_INCORRECT_STATE);
-    ReturnErrorOnFailure(EncodeSessionParameters(TLV::ContextTag(5), mLocalMRPConfig.Value(), tlvWriterMsg2));
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(ToRaw(Sigma2Tags::kResponderRandom)), &input.responderRandom[0],
+                                                sizeof(input.responderRandom)));
+    ReturnErrorOnFailure(tlvWriterMsg2.Put(TLV::ContextTag(ToRaw(Sigma2Tags::kResponderSessionId)), input.responderSessionId));
+
+    VerifyOrReturnError(input.pEphPubKey != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(ToRaw(Sigma2Tags::kResponderEphPubKey)), *input.pEphPubKey,
+                                                static_cast<uint32_t>(input.pEphPubKey->Length())));
+
+    // Check if msg_R2_Encrypted is not nullptr
+    VerifyOrReturnError(input.msg_R2_Encrypted, CHIP_ERROR_INCORRECT_STATE);
+
+    // Check if length of msg_R2_Encrypted is set and is at least larger than the MIC length
+    VerifyOrReturnError(input.encrypted2Length > CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(ToRaw(Sigma2Tags::kEncrypted2)), input.msg_R2_Encrypted.Get(),
```

Comment:
After this PutBytes, consider freeing the message buffer.

Reply:
It will be freed once we EncodeSigma2Inputs goes out of scope and we exit `HandleSigma1_and_SendSigma2`, wouldn't it? since it is a ScopedMemoryBuffer

Reply:
Yes, it will be.

I want implying that perhaps it should be scoped shorter.